WordPress/openverse-catalog
# Ensure SMK images don't timeout on validation

## Fixes


Fixes #494, #334

## Description


A little bit of background:

SMK (National Gallery of Denmark) returns both HQ and 'legacy' images. HQ images are IIIF enabled and have an `image_iiif_id` and IIIF image/thumbnail urls. Legacy images do not have this `iiif_id` field at all, and no IIIF links.

Our SMK provider script currently assumes that _all_ images are IIIF enabled, and uses an IIIF `image_url` with the size set to `max` (example URL: https://iip.smk.dk/iiif/jp2/KKSgb5781.tif.jp2/full/max/0/default.jpg). These images are very large and take a very long time to download.

This PR does a couple things:
* Renames "statensmuseum" to SMK for clarity
* Updates the `image_url` for HQ (IIIF-enabled) images to max `2048` px wide, which is what [Open SMK itself uses](https://open.smk.dk/artwork/image/KKSgb20980?q=*&page=0)
* Adds support for legacy images as well. Since these do not have IIIF links, we fallback to the `image_native` and `image_thumbnail` fields

Lots of really wonderful images coming from this provider :) I'm excited to get them back into Openverse!

## Additional Notes

Because legacy images don't have an `iiif_id`, I decided to use `id` as the foreign_identifier. According to API docs this is `the ID from collection DB`.

I chose not to use the `image_native` and `image_thumbnail` fields for HQ images as well as legacy ones, because:
* `image_native` for HQ images gives us a max size download link like this: https://api.smk.dk/api/v1/download/W3siaW1nX3VybCI6Imh0dHBzOi8vaWlwLnNtay5kay9paWlmL2pwMi9ra3MxMDM3OWMudGlmLnJlY29uc3RydWN0ZWQudGlmLmpwMi9mdWxsL2Z1bGwvMC9uYXRpdmUuanBnIiwicHVibGljX2RvbWFpbiI6dHJ1ZSwib2JqZWN0X251bWJlciI6IktLUzEwMzc5YyIsIm51bSI6IiJ9XQ==/KKS10379c.jpg
* `image_thumbnail` for HQ images gives us the same IIIF link we're currently using, but with size `1024` instead of `400`, so we get larger thumbnails than we're currently using. In either case it's unclear whether this field is being used at all.



## Testing Instructions


In your local Airflow, run the SMK provider script. Check the logs for the `pull_data` task, and wait until it has written several 100 lines from buffer to disk. We want to make sure we get some legacy images, so give it at least 30 seconds.

Find and download the TSV file that was loaded into storage through your MinIO Console (localhost:5011). In the TSV, the first column is `id`, the third is `image_url`, and the fourth is `thumbnail`.

* You should be able to see some legacy images in the TSV. These are spotted most easily by scrolling down the first column (id) and looking for an id that looks like `1170000790_object`. Its `image_url` and `thumbnail` values should **both** be non-IIIF thumbnail links, something like: "https://api.smk.dk/api/v1/thumbnail/5249f164-62b1-46e5-b099-d9ba9f596407.jpg"
* The other rows are all HQ images. Take a look at the generated `image_url` and see that the size is 2048. Example, `https://iip.smk.dk/iiif/jp2/kksgb20949.tif.jp2/full/!2048,/0/default.jpg` (specifically looking at the `full/!2048,/`, as opposed to `full/max/`)

You can also time a HEAD request for a few URLs of each type to verify the speed improvement. Example:
```
// The old URL format we were using (very slow)
$ time curl -I https://iip.smk.dk/iiif/jp2/KKSgb5781.tif.jp2/full/max/0/default.jpg

// New HQ URL format, restricts the image size (fast)
$ time curl -I https://iip.smk.dk/iiif/jp2/KKSgb5781.tif.jp2/full/!2048,/0/default.jpg

// A legacy image (fast)
$ time curl https://api.smk.dk/api/v1/thumbnail/5249f164-62b1-46e5-b099-d9ba9f596407.jpg

```

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
